keps/sig-storage/1698-generic-inline-volumes/README.md

```diff
@@ -0,0 +1,540 @@
+- The volume will be created using the normal storage provisioning
+  mechanism, without having to modify the driver or its deployment.
+- Storage capacity tracking can be enabled also for such inline
+  volumes.
+
+### Non-Goals
+
+- This will not replace CSI ephemeral inline volumes.
+- Inline volumes could also be kept alive after their pod terminates,
+  but that is only useful if some higher-level logic then takes care
+  of deletion. For now this is out of scope.
+
+## Proposal
+
+A new volume source will be introduced:
+
+```
+typedef InlineVolumeSource struct {
+    VolumeClaimTemplate PersistentVolumeClaim
+    ReadOnly            bool
+}
+```
+
+This mimics a `PersistentVolumeClaimVolumeSource`, except that it
+contains a `PersistentVolumeClaim` instead of referencing one by name.
+
+A new controller in `kube-controller-manager` is responsible for
+creating new PVCs for each such inline volume. It does that:
+- with a deterministic name that is a concatenation of pod name and
+  the `Volume.Name` of the volume,
+- in the namespace of the pod,
+- with the pod as owner.
+
+Kubernetes already prevents adding, removing or updating volumes in a
+pod and the ownership relationship ensures that volumes get deleted
+when the pod gets deleted, so the new controller only needs to take
+care of creating missing PVCs.
+
+When the [volume scheduling
+library](https://github.com/kubernetes/kubernetes/tree/v1.18.0/pkg/controller/volume/scheduling)
+inside the kube-scheduler or kubelet encounter such a volume inside a
+pod, they determine the name of the PVC and proceed as they currently
+do for a `PersistentVolumeClaimVolumeSource`.
+
+### User Stories
+
+#### Persistent Memory as DRAM replacement for memcached
+
+Recent releases of memcached added [support for using Persistent
+Memory](https://memcached.org/blog/persistent-memory/) (PMEM) instead
+of normal DRAM. When deploying memcached through one of the app
+controllers, `InlineVolumeSource` makes it possible to request a volume
+of a certain size from a CSI driver like
+[PMEM-CSI](https://github.com/intel/pmem-csi).
+
+#### Local LVM storage as scratch space
+
+Applications working with data sets that exceed the RAM size can
+request local storage with performance characteristics or size that is
+not met by the normal Kubernetes `EmptyDir` volumes. For example,
+[TopoLVM](https://github.com/cybozu-go/topolvm) was written for that
+purpose.
+
+### Risks and Mitigations
+
+Enabling this feature allows users to create PVCs indirectly if they can
+create pods, even if they do not have permission to create them
+directly. Cluster administrators must be made aware of this. If this
+does not fit their security model, they can disable the feature
+through the feature gate that will be added for the feature.
+
+Alternatively, a label on a namespace could be used to disable the
+feature just for that namespace.
+
+## Design Details
+
+### PVC meta data
```

Comment:
Related to above, please note that the new controller reports any errors on PVC creation as events on the Pod. This way, user can get for example quota errors.

Reply:
Added under implementation details, together with name collision.